src/main/java/io/stargate/sgv2/jsonapi/service/embedding/DataVectorizer.java

```diff
@@ -43,48 +44,62 @@ public DataVectorizer(EmbeddingService embeddingService, JsonNodeFactory nodeFac
+          if (document.has(DocumentConstants.Fields.VECTOR_EMBEDDING_FIELD)) {
+            throw new JsonApiException(ErrorCode.INVALID_USAGE_OF_VECTORIZE);
+          }
+          final JsonNode jsonNode =
+              document.get(DocumentConstants.Fields.VECTOR_EMBEDDING_TEXT_FIELD);
+          if (jsonNode.isNull()) {
+            ((ObjectNode) document)
+                .put(DocumentConstants.Fields.VECTOR_EMBEDDING_FIELD, (String) null);
+            continue;
+          }
+          if (!jsonNode.isTextual()) {
+            throw new JsonApiException(ErrorCode.SHRED_BAD_VECTORIZE_VALUE);
+          }
 
-        vectorizeTexts.add(jsonNode.asText());
-        vectorizeMap.put(vectorDataPosition, position);
-        vectorDataPosition++;
+          vectorizeTexts.add(jsonNode.asText());
+          vectorizeMap.put(vectorDataPosition, position);
+          vectorDataPosition++;
+        }
       }
-    }
 
-    if (!vectorizeTexts.isEmpty()) {
-      if (embeddingService == null) {
-        throw new JsonApiException(ErrorCode.UNAVAILABLE_EMBEDDING_SERVICE);
-      }
-      List<float[]> vectors = embeddingService.vectorize(vectorizeTexts);
-      for (int vectorPosition = 0; vectorPosition < vectors.size(); vectorPosition++) {
-        int position = vectorizeMap.get(vectorPosition);
-        JsonNode document = documents.get(position);
-        float[] vector = vectors.get(vectorPosition);
-        final ArrayNode arrayNode = nodeFactory.arrayNode(vector.length);
-        for (float listValue : vector) {
-          arrayNode.add(nodeFactory.numberNode(listValue));
+      if (!vectorizeTexts.isEmpty()) {
+        if (embeddingService == null) {
+          throw new JsonApiException(ErrorCode.UNAVAILABLE_EMBEDDING_SERVICE);
         }
-        ((ObjectNode) document).put(DocumentConstants.Fields.VECTOR_EMBEDDING_FIELD, arrayNode);
+        Uni<List<float[]>> vectors = embeddingService.vectorize(vectorizeTexts);
+        return vectors
+            .onItem()
+            .transform(
+                vectorData -> {
+                  for (int vectorPosition = 0;
+                      vectorPosition < vectorData.size();
+                      vectorPosition++) {
+                    int position = vectorizeMap.get(vectorPosition);
+                    JsonNode document = documents.get(position);
+                    float[] vector = vectorData.get(vectorPosition);
+                    final ArrayNode arrayNode = nodeFactory.arrayNode(vector.length);
+                    for (float listValue : vector) {
+                      arrayNode.add(nodeFactory.numberNode(listValue));
+                    }
+                    ((ObjectNode) document)
+                        .put(DocumentConstants.Fields.VECTOR_EMBEDDING_FIELD, arrayNode);
+                  }
+                  return true;
+                });
       }
+      return Uni.createFrom().item(true);
+    } catch (JsonApiException e) {
```

Comment:
I assume there's default handler in case some other exception was thrown (`NullPointerException` etc)